pkg/internal/ebpf/logenricher/logenricher.go

```diff
@@ -0,0 +1,316 @@
+		}
+	}
+
+	return newPids
+}
+
+func (p *Tracer) addPID(key uint64) error {
+	if err := p.bpfObjects.LogEnricherPids.Put(key, uint8(1)); err != nil {
+		return fmt.Errorf("error adding pid %d (ns=%d) to bpf map: %w", uint32(key), key>>32, err)
+	}
+	return nil
+}
+
+func (p *Tracer) removePID(key uint64) error {
+	if err := p.bpfObjects.LogEnricherPids.Delete(key); err != nil {
+		return fmt.Errorf("error removing pid %d (ns=%d) from bpf map: %w", uint32(key), key>>32, err)
+	}
+	return nil
+}
+
+func (p *Tracer) initPIDsMap(m *ebpf.Map) error {
+	if m == nil {
+		return errors.New("pids bpf map is nil")
+	}
+
+	pids := p.flattenPIDs()
+
+	p.log.Debug("allowing pids", "count", len(pids))
+
+	for _, pid := range pids {
+		if err := p.addPID(pid); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *Tracer) AllowPID(pid, ns uint32, svc *svc.Attrs) {
+	p.log.Debug("adding pid", "pid", pid, "ns", ns)
+	p.pf.AllowPID(pid, ns, svc, ebpfcommon.PIDTypeLogEnricher)
+	if err := p.addPID(p.pidKey(ns, pid)); err != nil {
+		p.log.Error(err.Error())
+	}
+}
+
+func (p *Tracer) BlockPID(pid, ns uint32) {
+	p.log.Debug("removing pid", "pid", pid, "ns", ns)
+	p.pf.BlockPID(pid, ns)
+	if err := p.removePID(p.pidKey(ns, pid)); err != nil {
+		p.log.Error(err.Error())
+	}
+}
+
+func (p *Tracer) Run(ctx context.Context, eventCtx *ebpfcommon.EBPFEventContext, _ *msg.Queue[[]request.Span]) {
+	p.log.Debug("starting")
+
+	if err := p.initPIDsMap(p.bpfObjects.LogEnricherPids); err != nil {
+		p.log.Error("failed to init pids map, not starting", "error", err)
+		return
+	}
+
+	ebpfcommon.ForwardRingbuf(
+		&p.cfg.EBPF,
+		p.bpfObjects.LogEvents,
+		eventCtx.CommonPIDsFilter,
+		p.handleLogEvent,
+		p.log,
+		nil,
+		nil,
+		append(p.closers, &p.bpfObjects)...,
+	)(ctx, nil)
+
+	p.log.Debug("terminating")
+}
+
+func (p *Tracer) Required() bool {
+	return false
+}
+
+func (p *Tracer) handleLogEvent(_ *ebpfcommon.EBPFParseContext, _ *config.EBPFTracer, record *ringbuf.Record, _ ebpfcommon.ServiceFilter) (request.Span, bool, error) {
+	hdrSize := uint32(unsafe.Sizeof(BpfLogEventT{})) - uint32(unsafe.Sizeof(uintptr(0))) // Remove `log` placeholder
+
+	event, err := ebpfcommon.ReinterpretCast[BpfLogEventT](record.RawSample)
+	if err != nil {
+		// This should never happen -- if it does, we can't really recover
+		// and the targeted process will miss his logs.
+		return request.Span{}, true, err
+	}
+
+	procFdPath := func(fd int) string {
+		return filepath.Join("/proc", strconv.FormatUint(uint64(event.Tgid), 10), "fd", strconv.Itoa(fd))
+	}
+
+	var filePath string
+	if event.Fd != 0 {
+		// This is a pipe write, use the target process pipe fd
+		filePath = procFdPath(int(event.Fd))
+	} else {
+		// TTY write
+		filePath = unix.ByteSliceToString(event.FilePath[:])
+		if filePath == "" {
+			// Fallback to process stdout in the case path resolver failed
+			filePath = procFdPath(1)
+		}
+	}
+
+	f, err := os.OpenFile(filePath, os.O_WRONLY, 0)
```

Comment:
Are we concerned this might be slow, opening and closing the file every time? Would it make sense to cache file descriptors with some sort of a timeout?

Reply:
makes sense, I can add a fd cache